pybotx_widgets/carousel.py

```diff
@@ -1,317 +1,324 @@
+            return False
 
+        return True
 
-def _is_valid_label_for_placing_numbers(label: str) -> bool:
-    number_places = list(string.Formatter().parse(label))
+    def _validate_params(self) -> None:
+        if "{selected_val}" not in self.SELECTED_VALUE_LABEL:
+            raise ValueError("'SELECTED_VALUE_LABEL' should contains '{selected_val}'")
 
-    if len([place for place in number_places if place[1] == ""]) != 2:
-        return False
+        if self._start_from > len(self.widget_content):
+            raise ValueError("'start_from' is greater than 'widget_content'")
 
-    return True
+        if self.loop and self.show_numbers:
+            raise ValueError("Sorry, you can't enable both 'loop' and 'show_numbers'")
 
+        if self.inline and self.show_numbers:
+            raise ValueError("Sorry, you can't enable both 'inline' and 'show_numbers'")
 
-def _get_inline_markup(
-    message: Message,
-    command: str,
-    displayed_content: Iterator,
-    show_left_arrow: bool,
-    show_right_arrow: bool,
-    control_labels: Tuple[str, str],
-    additional_markup: MessageMarkup = None,
-) -> MessageMarkup:
-    """Build inline markup for Carousel widget."""
-
-    markup = MessageMarkup()
-
-    if show_left_arrow:
-        markup.add_bubble(
-            command=message.command.command,
-            label=control_labels[0],
-            data={**message.data, "carousel_selected_val": LEFT_PRESSED},
-            new_row=False,
-        )
-
-    for content_item in displayed_content:
-        markup.add_bubble(
-            command=command,
-            label=content_item,
-            data={**message.data, "carousel_selected_val": content_item},
-            new_row=False,
-        )
-
-    if show_right_arrow:
-        markup.add_bubble(
-            command=message.command.command,
-            label=control_labels[1],
-            data={**message.data, "carousel_selected_val": RIGHT_PRESSED},
-            new_row=False,
-        )
-
-    if additional_markup:
-        markup = merge_markup(markup, additional_markup)
-
-    return markup
-
-
-def _get_newline_markup(
-    message: Message,
-    command: str,
-    displayed_content: Iterator,
-    show_left_arrow: bool,
-    show_right_arrow: bool,
-    show_numbers: bool,
-    pagination: PaginationInfo,
-    control_labels: Tuple[str, str],
-    additional_markup: MessageMarkup = None,
-) -> MessageMarkup:
-    """Build newline markup for Carousel widget."""
+        if not self.control_labels:
+            self.control_labels = (
+                (f"{self.LEFT_ARROW} ({{}}-{{}})", f"{self.RIGHT_ARROW} ({{}}-{{}})")
+                if self.show_numbers
+                else (self.LEFT_ARROW, self.RIGHT_ARROW)
+            )
 
-    markup = MessageMarkup()
-    right_label = control_labels[1]
+        if self.show_numbers:
+            if not self._is_valid_label_for_placing_numbers(self.control_labels[0]):
```

Comment:
Nit: you can use properties for left and right control labels.

Reply:
done